pkg/availability/multisigcollector/internal/parts/certverification/certverification.go

```diff
@@ -4,103 +4,179 @@ import (
-		if err != nil {
-			adsl.CertVerified(m, t.ModuleID(origin.Module), err, origin)
+	apbdsl.UponVerifyCert(m, func(cert *apbtypes.Cert, origin *apbtypes.VerifyCertOrigin) error {
+
+		reqID := state.NextReqID
+		state.NextReqID++
+
+		mscCerts, err := verifyCertificateStructure(params, cert)
+		valid, errStr := t.ErrorPb(err)
+
+		if err != nil { // invalid certificate, send cert verified event with valid=False
+			apbdsl.CertVerified(m, origin.Module, valid, errStr, origin)
 			return nil
 		}
 
-		sigMsg := common.SigData(params.InstanceUID, t.BatchID(mscCert.BatchId))
-		dsl.VerifyNodeSigs(m, mc.Crypto,
-			/*data*/ sliceutil.Repeat(sigMsg, len(mscCert.Signers)),
-			/*signatures*/ mscCert.Signatures,
-			/*nodeIDs*/ t.NodeIDSlice(mscCert.Signers),
-			/*context*/ &verifySigsInCertContext{origin},
-		)
+		state.RequestState[reqID] = &RequestState{
+			certVerifiedValid: make(map[t.BatchIDString]bool),
+		}
+		nonEmptyCerts := make([]*mscpbtypes.Cert, 0)
+
+		for _, mscCert := range mscCerts {
+			if !emptycert.IsEmpty(mscCert) {
+				nonEmptyCerts = append(nonEmptyCerts, mscCert) // do not add to requests
+			}
+		}
+		if len(nonEmptyCerts) > 0 {
+			for _, mscCert := range nonEmptyCerts {
+				state.RequestState[reqID].certVerifiedValid[t.BatchIDString(mscCert.BatchId)] = false
+				sigMsg := common.SigData(params.InstanceUID, mscCert.BatchId)
+				dsl.VerifyNodeSigs(m, mc.Crypto,
+					/*data*/ sliceutil.Repeat(sigMsg, len(mscCert.Signers)),
+					/*signatures*/ mscCert.Signatures,
+					/*nodeIDs*/ mscCert.Signers,
+					/*context*/ &verifySigsInCertContext{origin, reqID, mscCert},
+				)
+			}
+		} else {
+			// all certs are empty cert, verify immediately
+			apbdsl.CertVerified(m, origin.Module, true, "", origin)
+			delete(state.RequestState, reqID)
+		}
+
 		return nil
 	})
 
 	// When the signatures in a certificate are verified, output the result of certificate verification.
 	dsl.UponNodeSigsVerified(m, func(nodeIDs []t.NodeID, errs []error, allOK bool, context *verifySigsInCertContext) error {
+		reqID := context.reqID
+
+		if _, ok := state.RequestState[reqID]; !ok {
+			return nil
+		}
+
+		state.RequestState[reqID].certVerifiedValid[t.BatchIDString(string(context.cert.BatchId))] = allOK
 		var err error
 		if !allOK {
 			err = errors.New("some signatures are invalid")
+			valid, errStr := t.ErrorPb(err)
+			apbdsl.CertVerified(m, context.origin.Module, valid, errStr, context.origin)
+			delete(state.RequestState, reqID)
 		}
 
-		adsl.CertVerified(m, t.ModuleID(context.origin.Module), err, context.origin)
+		// if we get here, that means so far all received signatures for all certificates in the reqID are verified valid or yet to be verified
+		allVerifiedValid := true
+		for _, certVerifiedValid := range state.RequestState[reqID].certVerifiedValid {
+			if !certVerifiedValid {
+				allVerifiedValid = false
+				break
+			}
+		}
+
+		if _, ok := state.RequestState[reqID]; allVerifiedValid && ok {
```

Comment:
Haven't we already checked (on line 92) that the request is still pending?

Reply:
correct

Reply:
Then I suggest changing this condition to not check again here.

Reply:
done